### PR TITLE
AEJ 1625 - check added for input output buffer length and test cases

### PR DESCRIPTION
**Pull request by SnehalA**

### CODE REVIEW

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java [hunk 1]
```diff
@@ -199,8 +201,14 @@ public int deflate(byte[] b, int off, int len) {
         if(len <= 0) {
             throw new ArrayIndexOutOfBoundsException("Length value is less or equal than zero");
         }
+        if(inputBuffer == null || inputBufferLength == 0) {
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/compression/IntelDeflater.java:204]: Extra check for input buffer

File: src/main/java/com/intel/gkl/compression/IntelDeflater.java [hunk 2]
```diff
@@ -199,8 +201,14 @@ public int deflate(byte[] b, int off, int len) {
...
+            throw new NullPointerException("Input buffer is null");
+        }
 
-        return deflateNative(b, len);
+        int bytesWritten = deflateNative(b, len);
+        if(bytesWritten == 0)
+            logger.warn(String.format(" Zero Bytes Written : %d", bytesWritten));
```
Comment (SnehalA) [inline src/main/java/com/intel/gkl/compression/IntelDeflater.java:210]: warning for zero bytes written

File: src/main/native/compression/IntelDeflater.cc [hunk 3]
```diff
@@ -234,181 +234,194 @@ JNIEXPORT jint JNICALL Java_com_intel_gkl_compression_IntelDeflater_deflateNativ
   //jint inputBufferOffset = env->GetIntField(obj, FID_inputBufferOffset);
   jboolean endOfStream = env->GetBooleanField(obj, FID_endOfStream);
   jint level = env->GetIntField(obj, FID_level);
+  const char* err_msg;
+
+  if( outputBufferLength == 0 || inputBufferLength == 0 )
```
Comment (SnehalA) [inline src/main/native/compression/IntelDeflater.cc:239]: Precheck for deflate

File: src/main/native/compression/IntelDeflater.cc [hunk 4]
```diff
@@ -234,181 +234,194 @@ JNIEXPORT jint JNICALL Java_com_intel_gkl_compression_IntelDeflater_deflateNativ
...
-      return -1;
-    }
-
-    lz_stream->next_in = (uint8_t*) (next_in);
-    lz_stream->avail_in = inputBufferLength;
-    lz_stream->end_of_stream = endOfStream;
-    lz_stream->next_out = (uint8_t*) (next_out);
-    lz_stream->avail_out = outputBufferLength ;
-
-    int bytes_in = inputBufferLength;
```
Comment (SnehalA) [inline src/main/native/compression/IntelDeflater.cc:269]: removed this variable as not used.

File: src/main/native/compression/IntelDeflater.cc [hunk 5]
```diff
@@ -234,181 +234,194 @@ JNIEXPORT jint JNICALL Java_com_intel_gkl_compression_IntelDeflater_deflateNativ
...
+          const char* err_msg;
+          switch (ret) {
+            case Z_STREAM_ERROR:
+              err_msg = "Stream state is inconsistent.";
+              break;
+            case Z_BUF_ERROR:
+              err_msg = "No progress possible. Buffer error.";
+              break;
+            default:
+              err_msg = zError(ret);
```
Comment (SnehalA) [inline src/main/native/compression/IntelDeflater.cc:405]: Added ref to zError than generic msg

File: src/test/java/com/intel/gkl/compression/DeflaterUnitTest.java [hunk 6]
```diff
@@ -148,6 +147,46 @@ public void deflateThrowsArrayIndexOutOfBoundsExceptionWhenLengthEqualZeroTest()
...
 
+    @Test(enabled = true, expectedExceptions = RuntimeException.class)
+    public void deflateEmptyOutputBufferTest(){
+        final Deflater deflater = new IntelDeflaterFactory().makeDeflater(0, true);
+        int LEN = 10;
+        byte[] inputBuffer = new byte[LEN];
+        TestingUtils.randomDNA(inputBuffer);
+
+        int bytes = 0;
+        byte[] outputBufferBad = new byte[0];
```
Comment (SnehalA) [inline src/test/java/com/intel/gkl/compression/DeflaterUnitTest.java:158]: not null but size 0

File: src/test/java/com/intel/gkl/compression/InflaterUnitTest.java [hunk 7]
```diff
@@ -88,6 +88,112 @@ public void inflateThreeArgsThrowsNullPointerExceptionWhenNullBufferTest() throw
...
+        final Inflater inflater = new IntelInflaterFactory().makeInflater(nowrap);
+        final IntelDeflaterFactory intelDeflaterFactory = new IntelDeflaterFactory();
+        final Deflater deflater = intelDeflaterFactory.makeDeflater(level, nowrap);
+
+        try
+        {
+            String sequence = "ACTGTC";
+            byte[] input = sequence.getBytes("UTF-8");
+            byte[] compressed = new byte[sequence.length()];
+            byte[] result = new byte[sequence.length() - 2];
```
Comment (SnehalA) [inline src/test/java/com/intel/gkl/compression/InflaterUnitTest.java:119]: Prints ERROR InflaterUnitTest - Output buffer too small.

**Review by Kmannth — APPROVED** [review]

**Pull request merged**